support/cas-server-support-simple-mfa-core/src/main/java/org/apereo/cas/mfa/simple/validation/DefaultCasSimpleMultifactorAuthenticationService.java

```diff
@@ -38,9 +41,15 @@ public DefaultCasSimpleMultifactorAuthenticationService(final TicketRegistry tic
     @Override
     public CasSimpleMultifactorAuthenticationTicket generate(final Principal principal, final Service service) throws Throwable {
         val mfaFactory = (CasSimpleMultifactorAuthenticationTicketFactory) ticketFactory.get(CasSimpleMultifactorAuthenticationTicket.class);
-        val token = mfaFactory.create(service, CollectionUtils.wrap(CasSimpleMultifactorAuthenticationConstants.PROPERTY_PRINCIPAL, principal));
-        LOGGER.debug("Created multifactor authentication token [{}] for service [{}]", token.getId(), service);
-        return token;
+        for (var i = 0; i < MAX_ATTEMPTS; i++) {
+            val token = mfaFactory.create(service, CollectionUtils.wrap(CasSimpleMultifactorAuthenticationConstants.PROPERTY_PRINCIPAL, principal));
+            val trackingToken = ticketRegistry.getTicket(token.getId());
```

Comment:
Also this line should likely be reworked, because it will throw an exception. This is an artifact of using mocks that hide such things. Avoid mocks if you can.

Reply:
You should consider using `doAndRetry`, which has a configurable number of attempts, and also supports delays between each op. Brute forcing the number generation in a loop leaves very little chance to a code to free up or entropy to generate the next best random number. You also "throttle" the generation process a bit to control the flow.

Reply:
About the "it will throw an exception", I don't think so, it's the `getTicket(final String ticketId, final @NonNull Class<T> clazz)` method which throws `Exception`, not the `getTicket(final String ticketId)` method (it only cleans expired ticket).

If so, as it is used in the `store` method of the `DefaultCasSimpleMultifactorAuthenticationService` component, it would always throw an `Exception` and this whole mechanism would never work.

But you're right about the mock drawback here, I will try to use a real ticket registry as I should be able to store the tickets I want and at the same time confirm no exception is thrown.